src/System.Management.Automation/DscSupport/JsonDscClassCache.cs

```diff
@@ -0,0 +1,3249 @@
+    /// </summary>
+    internal class DscClassCacheEntry
+    {
+        /// <summary>
+        /// Store the RunAs Credentials that this DSC resource will use.
+        /// </summary>
+        public DSCResourceRunAsCredential DscResRunAsCred { get; set; }
+
+        /// <summary>
+        /// If we have implicitly imported this resource, we will set this field to true. This will
+        /// only happen to InBox resources.
+        /// </summary>
+        public bool IsImportedImplicitly { get; set; }
+
+        /// <summary>
+        /// A CimClass instance for this resource.
+        /// </summary>
+        public PSObject CimClassInstance { get; set; }
+
+        /// <summary>
+        /// Initializes variables with default values.
+        /// </summary>
+        public DscClassCacheEntry() : this(DSCResourceRunAsCredential.Default, false, null) { }
+
+        /// <summary>
+        /// Initializes all values.
+        /// </summary>
+        /// <param name="aDSCResourceRunAsCredential"></param>
+        /// <param name="aIsImportedImplicitly"></param>
+        /// <param name="aCimClassInstance"></param>
+        public DscClassCacheEntry(DSCResourceRunAsCredential aDSCResourceRunAsCredential, bool aIsImportedImplicitly, PSObject aCimClassInstance)
+        {
+            DscResRunAsCred = aDSCResourceRunAsCredential;
+            IsImportedImplicitly = aIsImportedImplicitly;
+            CimClassInstance = aCimClassInstance;
+        }
+    }
+
+    /// <summary>
+    /// </summary>
+    [SuppressMessage("Microsoft.MSInternal", "CA903:InternalNamespaceShouldNotContainPublicTypes",
+        Justification = "Needed Internal use only")]
+    public static class DscClassCache
+    {
+        private const string windowsInboxDscResourceModulePath = "WindowsPowershell\\v1.0\\Modules\\PsDesiredStateConfiguration";
+
+        private const string reservedDynamicKeywords = "^(Synchronization|Certificate|IIS|SQL)$";
+
+        private const string reservedProperties = "^(Require|Trigger|Notify|Before|After|Subscribe)$";
+
+        private const string jsonSchemaSupportExperimentalFeatureName = "PSDscJsonSchemaSupport";
+
+        private static readonly PSTraceSource s_tracer = PSTraceSource.GetTracer("DSC", "DSC Class Cache");
+
+        // Constants for items in the module qualified name (Module\Version\ClassName)
+        private const int IndexModuleName = 0;
+        private const int IndexModuleVersion = 1;
+        private const int IndexClassName = 2;
+        private const int IndexFriendlyName = 3;
+
+        // Create a list of classes which are not actual DSC resources similar to what we do inside PSDesiredStateConfiguration.psm1
+        private static readonly string[] s_hiddenResourceList =
+        {
+            "MSFT_BaseConfigurationProviderRegistration",
+            "MSFT_CimConfigurationProviderRegistration",
+            "MSFT_PSConfigurationProviderRegistration",
+        };
+
```

Comment:
Below a `HashSet<string>` is instantiated with its arguments directly, rather than needing the intermediate array. Perhaps the same could be done here?

Reply:
updated.